crates/revm/src/instructions/bitwise.rs

```diff
@@ -1,80 +1,135 @@
+    pop_top!(interpreter, op1, op2);
+    *op2 = if op1.gt(op2) {
         U256::from(1)
     } else {
         U256::ZERO
-    }
+    };
 }
 
-pub fn iszero(op1: U256) -> U256 {
-    if op1 == U256::ZERO {
+pub fn slt(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    pop_top!(interpreter, op1, op2);
+    *op2 = if i256_cmp(op1, *op2) == Ordering::Less {
         U256::from(1)
     } else {
         U256::ZERO
     }
 }
 
-pub fn not(op1: U256) -> U256 {
-    !op1
+pub fn sgt(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    pop_top!(interpreter, op1, op2);
+    *op2 = if i256_cmp(op1, *op2) == Ordering::Greater {
+        U256::from(1)
+    } else {
+        U256::ZERO
+    };
+}
+
+pub fn eq(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    pop_top!(interpreter, op1, op2);
+    *op2 = if op1.eq(op2) {
+        U256::from(1)
+    } else {
+        U256::ZERO
+    };
+}
+
+pub fn iszero(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    pop_top!(interpreter, op1);
+    *op1 = if *op1 == U256::ZERO {
+        U256::from(1)
+    } else {
+        U256::ZERO
+    };
+}
+pub fn bitand(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    pop_top!(interpreter, op1, op2);
+    *op2 = op1.bitand(*op2);
+}
+pub fn bitor(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    pop_top!(interpreter, op1, op2);
+    *op2 = op1.bitor(*op2);
+}
+pub fn bitxor(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    pop_top!(interpreter, op1, op2);
+    *op2 = op1.bitxor(*op2);
+}
+
+pub fn not(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    pop_top!(interpreter, op1);
+    *op1 = !*op1;
 }
 
-pub fn byte(op1: U256, op2: U256) -> U256 {
+pub fn byte(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    pop_top!(interpreter, op1, op2);
     let mut ret = U256::ZERO;
 
     for i in 0..256 {
         if i < 8 && op1 < U256::from(32) {
-            let o = u128::try_from(op1).unwrap() as usize;
+            let o = as_usize_saturated!(op1);
             let t = 255 - (7 - i + 8 * o);
             let bit_mask = U256::from(1) << t;
-            let value = (op2 & bit_mask) >> t;
+            let value = (*op2 & bit_mask) >> t;
             ret = ret.overflowing_add(value << i).0;
         }
     }
 
-    ret
+    *op2 = ret;
 }
 
-pub fn shl(shift: U256, value: U256) -> U256 {
-    value << usize::try_from(shift).unwrap_or(256)
+pub fn shl<SPEC: Spec>(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    // EIP-145: Bitwise shifting instructions in EVM
+    check!(interpreter, SPEC::enabled(CONSTANTINOPLE));
+    pop_top!(interpreter, op1, op2);
+    *op2 <<= as_usize_saturated!(op1);
 }
 
-pub fn shr(shift: U256, value: U256) -> U256 {
-    value >> usize::try_from(shift).unwrap_or(256)
+pub fn shr<SPEC: Spec>(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    // EIP-145: Bitwise shifting instructions in EVM
+    check!(interpreter, SPEC::enabled(CONSTANTINOPLE));
```

Comment:
Can we do a static analysis pass at the bytecode in the beginning for the existence of these opcodes against the spec, instead of doing them in the loop?

Reply:
Will think about it, it is small thing as branch prediction does a great job, but will have this idea in my head, it does make sense.